12.0 theme seb

seb-odoo opened the pull request:
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

qsm-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -5,143 +5,117 @@
+        'website.page': 'theme.website.page',
+        'website.menu': 'theme.website.menu',
+        'ir.attachment': 'theme.ir.attachment',
+    }
```
dict order is not preserved

seb-odoo replied:
changed to ordereddict

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -449,14 +449,24 @@ def get_url_localized(router, lang):
             return self.browse(website_id)
 
+        if only_explicit:
+            return False
```
Maybe the solution is to return the first website in this case

seb-odoo replied:
I removed only_explicit and used "if request" to differentiate with -u. Much better now

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -5,143 +5,116 @@
                     find = rec.copy_ids.search([('key', '=', rec.key), ('website_id', '=', website.id)])
 
-                if old and model == 'ir.ui.view': # at update, ignore active field
+                if find and 'active' in rec_data:  # at update, ignore active field
```
Done, I reorganized the rest of that if/else in the same spirit.

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
+        if len(remaining):
+            error = 'Error - Remaining: %s' % remaining.mapped('display_name')
+            _logger.error(error)
+            raise MissingError(error)
```
It will log something when called from command -u but it will not log anything when called from the backend UI.

JKE-be replied:
chez moi ca logue bien les 2, -u ou update

seb-odoo replied:
I'll try one last time to be sure then.

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
-            mod._unload_one_theme_module(website)
+        orphans = self._theme_find_orphans(model_name, website)
+        orphans.unlink()
+        return orphans
```
unlink always return true.

I don't actually need to return anything, but the original method returned the created,updated,deleted models.

I don't know if I need to keep that feature or not, because those variables are then never used.

JKE-be replied:
or you return ids delete or nothing, but not true ;)

seb-odoo replied:
Is it ok if _update_records also return nothing then?

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
-            _logger.info('Load theme %s for website %s from template.' % (mod.name, website.id))
-            mod._load_one_theme_module(website, with_update=False)
-            self.env['theme.utils']._post_copy(mod)
+        return self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
```
Reversed is needed only on "_remove_theme" to remove them on the opposite order they were added.

Reversed is not needed anymore in the other cases because the different "_theme_get_stream_themes" methods return them in the correct order on purpose.

As for the upgrade is triggered automatically in the write, so it will always be in the correct order because that's how modules are upgraded in general.

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
+        self.ensure_one()
+        websites = self.env['website']
+        for website in self.env['website'].search([('theme_id', '!=', False)]):
+            for theme in website.theme_id._theme_get_downstream():
```
I changed it with 'if self in'

seb-odoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
+        websites = self.env['website'].search([('theme_id', '!=', False)])
+        for website in websites:
+            for theme in website.theme_id._theme_get_stream_themes():
+                if self == theme:
```
I changed it with 'if self in'

rdeodoo commented:
Everything seems to **works as it should** except as discussed, updating a theme will only update its direct upstream on one level and not all the upstream (eg: Updating Bistro will update Treehouse but not Common):

============ -u in console ============
Updating a theme with `-u` in console (eg: `-u theme_treehouse`) will update that theme, load that theme on website having that theme loaded (or a downstream of that theme).
Then it will do the same flow for that theme's downstream themes (Update the theme, then load it)
It will also update dependency themes (theme_*_blog, theme_*_sale etc)

So having:
Website 1: Avantgarde (> Common)
Website 2: Bistro (> Treehouse > Common)
Website 3: Treehouse (> Common)
And website_blog installed

Doing a `-u theme_common` will:
1. Update theme_common module
2. Load theme_common on website 1
3. Load theme_common on website 2
4. Load theme_common on website 3
5. Update theme_avantgarde module
6. Load theme_avantgarde on website 1
7. Update theme_treehouse module
8. Load theme_treehouse on website 2
9. Load theme_treehouse on website 3
10. Update theme_avantgarde_blog module
11. Load theme_avantgarde_blog on website 1
12. update theme_bistro module
13. Load theme_bistro on website 2

Doing a `-u theme_treehouse` will:
1. Update theme_treehouse module
2. Load theme_treehouse on website 2
3. Load theme_treehouse on website 3
4. update theme_bistro module
5. Load theme_bistro on website 2

============ update theme in backend ============
Updating a theme in backend from website settings "Choose a theme" will update the upstream themes starting from the higher one, then load that theme on the selected website.
It will then update all this theme downstream (potentially from other websites, without loading them).
It will then update other upstream themes by descending the hierarchy, doing the same flow (Updating, loading, updating downstream without loading)

So having:
Website 1: Avantgarde (> Common)
Website 2: Bistro (> Treehouse > Common)
Website 3: Treehouse (> Common)
And website_blog installed

Updating Treehouse from website 3 settings will:
1. Update theme_common module
2. Load theme_common on website 3
3. Update theme_avantgarde module (As it is a downstream of theme_common, but won't load anything)
4. Update theme_treehouse module
5. Load theme_treehouse on website 3
6. Update theme_avantgarde_blog module (As it is a dependency of theme_avantgarde, but won't load anything)
7. Update theme_bistro module (As it is a downstream of theme_treehouse, but won't load anything)

**Can/should steps 3. 6. and 7. be avoided? As they are not loaded it does no harm?**

Updating Avantgarde from website 1 settings will:
1. Update theme_common module
2. Load theme_common on website 1
3. Update theme_avantgarde module
4. Load theme_avantgarde on website 1
5. Update theme_treehouse module (As it is a dependency of theme_common, but won't load anything)
6. Update theme_avantgarde_blog module
7. Load theme_avantgarde_blog on website 1
8. Update theme_bistro module (As it is a dependency of theme_treehouse, but won't load anything)

Updating Bistro from website 2 settings will:
1. Update theme_treehouse module
2. Load theme_treehouse on website 2
3. Update theme_bistro module
4. Load theme_bistro on website 2

**=> This is not correct as it should have firstly Update and Load theme_common**

============ Theme's blog/shop bridges ============
theme_avantgarde_blog will be installed automatically if both his dependencies are installed (website_blog and theme_avantgarde).
Both ways will do it correctly:
1. Install website_blog then theme_avantgarde, theme_avantgarde_blog will be installed/updated
2. Install theme_avantgarde then website_blog, theme_avantgarde_blog will be installed/updated

There is a lot of flows (I am probably missing a few), this PR need to be merged ASAP.
But once merged, it seems mandatory to write as much tests as possible (What about -u tests, possible?) as it is most likely that one of the flow will break on the first change we do later without even noticing it.

Sorry for the big message but I think this could be helpfull to write tests later

seb-odoo commented on addons/website/models/website.py:
```diff
@@ -443,14 +443,24 @@ def get_url_localized(router, lang):
+        """
+            Return the current website.
+
+            If ``only_explicit`` is set, do not try to guess,
```
I removed only_explicit and used "if request" to differentiate with -u. Much better now

seb-odoo commented:
I'm happy with the current state of the feature and code.

There is one functional part I'm not sure about, it's the upgrade of themes already used in websites, because both solutions can lead to breaking changes.

- If no reload is made after upgrade, we won't break views, but it might still go wrong if for example a JS or CSS file that is already linked has been changed. (current state of the code)

- But if we decide to always reload on upgrade, it might break/change the look and/or custo of websites that the user maybe didn't want to upgrade...

The solution would be to also allow "theme templates" for assets, and then to keep the "no upgrade" policy.

Ideally we need to write tests.

seb-odoo commented:
I did the last small fixes, added a lot of comments in the code, and squashed the commits into one with a detailed message.

Please make the final functional test and technical review.

seb-odoo commented:
I added a commit with my changes following the feedback (also updated the original commit message to reflect the changes made in the review).

Those changes are mostly adding comments, and "reverting" 3 small changes  (log error when request is used, unlink in cleanup and use the python way of "if else" instead of "and or")
.
Ok to remove the "keep_theme" for this PR.

I propose however that we create a "post_delete" doing the reverse of "post_copy" to handle this problem.

JKE-be commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -80,7 +81,7 @@ def write(self, vals):
                         Website = self.env['website']
                         current_website = Website.get_current_website()
-                        websites_to_update = current_website in websites_to_update and current_website or Website
+                        websites_to_update = current_website if current_website in websites_to_update else Website
```
NO DEBATE ABOUT IT !

Just for your information: https://www.pythoncentral.io/one-line-if-statement-in-python-ternary-conditional-operator/

seb-odoo replied:
Thank for the link!

I just want this merged so I'm not going to debate it and I'll change it back.

That said in the link it specifies "and or" to be better to not evaluate twice the expression, which I agree.

But in this case, this it's just reading a variable so there is no performance impact and it is more readable with "if else"...

seb-odoo commented:
merged 96ef4885a79757d038e3d152ac913fdd0b60e714
